### ISSUE TITLE
unable to register allowed origin

### ISSUE THREAD

**Issue by minhster99** [teslamotors/vehicle-command#159]
Hi, I'm attempting to register a domain I own https://robee.click but it says 
> Domain must be registered with a certificate authority. 

What does this mean exactly? It's hosted on AWS and has a valid certificate.

![image](https://github.com/teslamotors/vehicle-command/assets/36299836/dcc11276-10c9-4f1f-85e4-e7f245573319)

**Comment by derekargueta** [comment]
Checking this internally, may be that the Amazon root CA is not in the list of checked authorities.

**Comment by jbanyer** [comment]
@derekargueta I have successfully registered an app and configured Allowed Origin URLs including `https://app.chqtest.net`, which uses an Amazon CA cert. This was done some time ago.

I just tried editing the list of Allowed Origin URLs to add `https://api.chqtest.net`, and it worked.

Does this CA restriction only apply when creating a new app?

![Screenshot 2024-02-02 at 8 19 11 am](https://github.com/teslamotors/vehicle-command/assets/936491/dec1a456-5b02-4dbd-a366-245ef00b12ab)

**Comment by minhster99** [comment]
@jbanyer my https://robee.click site is failing on update. I just attempted your site and it worked. I checked the root CA on yours vs mine and it is slightly different. The CN on mine is `Amazon RSA 2048 M02` , and yours is `Amazon RSA 2048 M01`

@derekargueta any chance we could add the newer AWS root cert as trusted too?

**Comment by jbanyer** [comment]
@derekargueta I'll add my 2c: it would be very inconvenient if Amazon CAs are not accepted. AWS provides a very frictionless mechanism for adding and rotating certs when using their services, and I'm guessing a large number of Tesla API customers are going to use them by default. Cheers.

**Comment by llamafilm** [comment]
FWIW, I got the exact same error message for a different reason:  My webserver was listening for HTTPS on a nonstandard port.  When I changed it to 443 this error went away.

**Comment by DaveTBlake** [comment]
Having difficulty getting with my SSL, perhaps little wonder since the CA is not on that short list of acceptable providers. But is domain validation enough, or is Tesla expecting organisation validation level certification too?

**Comment by minhster99** [comment]
Does anyone know which of the acceptable CAs provide free SSL certs? This is a personal project and not sure I am willing to spend a few hundred bucks per year.

**Comment by llamafilm** [comment]
Lets Encrypt works for me

**Comment by minhster99** [comment]
@llamafilm I just tried LetsEncrypt but I'm still getting the invalid domain error. Is there something on the Tesla side caching the old cert perhaps?

<img width="546" alt="Screenshot 2024-02-12 at 12 26 59 pm" src="https://github.com/teslamotors/vehicle-command/assets/36299836/313eaaec-7412-4db8-b199-52d1139b430c">

**Comment by llamafilm** [comment]
Not sure.  Maybe they don't allow the .click TLD ?

**Comment by minhster99** [comment]
@llamafilm can you give me an example site you got working with LetsEncrypt? I'll compare the certs to see if there's a difference.

.click domain happened to be the cheapest which is why I got it.

**Comment by llamafilm** [comment]
Mine is working with https://tesla.llamafilm.com

**Comment by minhster99** [comment]
I've got something working. The answer is to have a .com domain and this will work with the AWS signed certificate as well.

@derekargueta I think you should improve the error message because it is currently misleading. Although I don't know why you wouldn't accept any domain as long as it exists. fyi .com domain costs about 4x more than .click domains

**Comment by paky79** [comment]
I have the same issue here because at the moment I'm running HA with Duckdns+nginx and duckdns is not recognized as allowed origin Does it mean that I should buy .com DNS and move everything in cloudflare? I saw that Tesla proxy requires anyway nginx. How to solve?

**Comment by handya** [comment]
I'm having the same issue on update, using a .app domain and cloudflare for the cert. I added it a few weeks ago and it worked no problem now when I'm trying to update I get "Domain must be registered with a certificate authority". 

I thought it was an issue with .app as others have mentioned but I tried my other .app domains and they worked, not sure why it's suddenly not liking my existing domain.

**Comment by getget1980** [comment]
Hello, same problem here, self-hosted domain, with .fr domain name, and using GeoTrust - powered by DigiCert certificate.

Certificate is delivered by Encryption Everywhere DV TLS CA - G2 which seems to be on the VISA list ...

**Comment by kiberxx** [comment]
Hello, facing the same error with "com" TLD and LetsEncrypt certificate. Does anyone know the trick how to make it work? On top of the comments above it is not clear what exactly is the problem. 
https://vdubinskii.mooo.com/

**Comment by getget1980** [comment]
It worked with Let's encrypt, so for me the problem was GeoTrust - powered by DigiCert certificate.

**Comment by kiberxx** [comment]
Yeah, exactly, checked the domain mentioned above with letsEncrypt cert and don't see a significant difference. They use wildcard cert and have a CNAME record, no CAA records. Weird, maybe it is due to I am using freedns. I don't want to play with the paid domains until the issue is clear.

**Comment by KevM** [comment]
I just ran into this issue today. My domain (hosted by Vercel) has a valid cert issued by Let's Encrypt.

**Comment by Sebazzz** [comment]
Also running into this issue, certificate issued by Let's Encrypt E6, with ISRG Root X1 as root.

**Comment by jmcollin78** [comment]
Hello same problem here with a domain in .fr and a Let's Encrypt certificate:

<img width="212" alt="Capture d’écran 2024-06-11 à 07 42 48" src="https://github.com/teslamotors/vehicle-command/assets/1712374/e9c927d7-2bb0-4be2-9d90-4f603a8ba7ef">

**Comment by getget1980** [comment]
Today it's not working anymore (without any changes)... I'm starting to being bored by all this stiff ....

**Comment by KevM** [comment]
My plan is to email support I’ll be needing this down the road.

**Comment by jmcollin78** [comment]
I wrote an email to the support at this address: fleetapisupport@tesla.com
I think you can also do the same to hope thinks will change.
I don't see any reason to refuse domain if signed with a valid certificate and root ca.

**Comment by sethterashima** [comment]
This scope of this issue tracker is limited to the Golang code in this repository. As @jmcollin78 suggests, please reach out to the Fleet API support team: https://developer.tesla.com/docs/fleet-api#help-and-support.

**Comment by KevM** [comment]
Reached out a week ago and still have not heard from them.

**Comment by jmcollin78** [comment]
No news also.

**Comment by ThomasAlxDmy** [comment]
Hey All,

We officially added support for Let's encrypt. I've checked some of the URL in this issue most certs are expired. Please renew them and try again. 

If you are having issue with Let's encrypt and your certs are valid add your URL to this issue so we can debug and see what's going on. Thank you!

**Comment by minhster99** [comment]
Hi @ThomasAlxDmy 

I am the original poster of this issue. Here are my findings 

| URL                 | CA          | Working |
|---------------------|-------------|---------|
| robee.click         | AWS         | No      |
| robee.click         | LetsEncrypt | No      |
| tesla.llamafilm.com | LetsEncrpt  | Yes     |
| robee2.com          | AWS         | Yes     |

Based on the above, both AWS and LetsEncrypt already works however it seems to require a .com domain. If you relax this requirement, it would likely solve a whole bunch of issues for everyone else here.

**Comment by KevM** [comment]
What’s weird is on the past I’ve registered a non .com domain without issue. (.rodeo for testing)

**Comment by ThomasAlxDmy** [comment]
Yup if set up properly it should work now. Most common issues we've seen:

1. certs are expired.
2. Common Name (CN) in the certificate not matching the URL

**Comment by firefixmaarten** [comment]
Hi, how long does it take for the certificate to be rechecked ? As my previous CN was using a wildcard.
It could also be that the dns propagation did not reach your end.

URL: https://tesla.maartenvandenbranden.be/

**Comment by timcantryn** [comment]
I can confirm it's not working on a .be domain with unexpired cert with correct CN, is it possible to get a confirmation whether the allowed domains are limited to the .com TLD or if there are certain unallowed TLD's?
Thanks!
![image](https://github.com/teslamotors/vehicle-command/assets/66787496/3d6dc6d0-50c0-4dae-a84c-5ac40f95159d)
![image](https://github.com/teslamotors/vehicle-command/assets/66787496/9abc143a-76ad-4073-a43b-7661363f3380)

**Comment by jmcollin78** [comment]
Hello, great new I can confirm it is now working with my .fr domain. Thanks for the fix

**Comment by firefixmaarten** [comment]
> I can confirm it's not working on a .be domain with unexpired cert with correct CN, is it possible to get a confirmation whether the allowed domains are limited to the .com TLD or if there are certain unallowed TLD's?
> Thanks!
> ![image](https://github.com/teslamotors/vehicle-command/assets/66787496/3d6dc6d0-50c0-4dae-a84c-5ac40f95159d)
> ![image](https://github.com/teslamotors/vehicle-command/assets/66787496/9abc143a-76ad-4073-a43b-7661363f3380)
> 

It works with tesla.maartenvandenbranden.be. I just got It working. Let's encrypt works ;).

**Comment by Jopand** [comment]
It is NOT working on valid *.synology.me, *.mywire.org and *.duckdns.org domains 🤔

**Comment by acanturgut** [comment]
My firebase hosted site not pass the validation as well. It looks super random my opinion. 

<img width="515" alt="Screenshot 2024-06-30 at 20 01 46" src="https://github.com/teslamotors/vehicle-command/assets/19980760/3b1e12fd-384b-401f-b20a-4a3524979c00">

**Comment by timcantryn** [comment]
In my case I still get the error on beforementioned .be domain, it does seem to be quite random.

**Comment by acanturgut** [comment]
Contacting with Tesla fix our problem. They added it manually behalf of us.

**Comment by sgryphon** [comment]
They seem to have some limitations on which domains they accept (even if the domains are valid, have valid certs, etc.

BTW. The best cert checker I found was Qualys/SSL Labs, i.e. https://www.ssllabs.com/ssltest/analyze.html?d=tesla.gamertheory.net

All the following had Let's Encrypt certs, but only 'www' worked.

* `tesla.gamertheory.net` -- didn't work, note that the host is in the SAN (subject alternative name), not the CN (common name)
* `tesla-auth.gamertheory.net` -- didn't work, for this on the host was the common name
* `www.gamertheory.net` -- did work

**Comment by sethterashima** [comment]
@timcantryn I see https://tesla.cantryn.be/ serves a certificate for localhost.localdomain, not tesla.cantryn.be.

@jopand can you provide specific subdomains to test?

@sgryphon `tesla-auth.gamertheory.net` isn't responding for HTTPS connections, I can only connect over HTTP. I'll follow up on SAN vs CN for `tesla.gamertheory.net`, I think that might one cause of the issue.

**Comment by timcantryn** [comment]
@sethterashima, what do you mean? This looks fine to me:
![IMG_5957](https://github.com/teslamotors/vehicle-command/assets/66787496/afbdb309-db5b-4e68-8ac7-613b66f31aca)

**Comment by sethterashima** [comment]
@timcantryn My mistake. Looks like IT doesn't like me visiting tesla.* domains that don't belong to them. :)

**Comment by sgryphon** [comment]
> @sgryphon tesla-auth.gamertheory.net isn't responding for HTTPS connections, I can only connect over HTTP. I'll follow up on SAN vs CN for tesla.gamertheory.net, I think that might one cause of the issue.

Well, it does respond to HTTPS for half the internet (see 'Before' result below), but I forgot about legacy internet, and didn't set up the reverse proxy for the new host name, so was only working for IPv6.

Before:
<img width="496" alt="image" src="https://github.com/teslamotors/vehicle-command/assets/339293/a5f67a97-8962-4fee-9bad-cc442c9d971b">

I have now added tesla-auth to reverse proxy, so legacy connections are possible:

After (now):
![image](https://github.com/teslamotors/vehicle-command/assets/339293/43c8c1d8-1d80-4658-a65a-9973937af98d)

Note that tesla.gamertheory.net is only IPv6, so you do need a modern internet connection to be able to access it. If a work computer is failing, try a mobile phone (or home Internet) ... or just use a checker like Qualys.

![image](https://github.com/teslamotors/vehicle-command/assets/339293/ca17a903-b2c2-4122-8835-6ae122f08085)

**Comment by martinzitka** [comment]
I have a similar problem to @sgryphon above: my domain is still being rejected

![image](https://github.com/teslamotors/vehicle-command/assets/4542639/96fa501d-3316-4dcd-abe2-bbea5707341b)

despite getting an A+ SSL rating:

![image](https://github.com/teslamotors/vehicle-command/assets/4542639/bbad1f5f-d1f5-4894-bd28-9b89447f13b1)

And yes, the only reason I can think of is the fact I am running the service on Internet and not legacy internet. Honestly I had no intent running this on IPv4.

Could someone from the Tesla team please confirm that their service is currently only supported on legacy IPv4, and if so, if there are any plans to also support IPv6? Thank you!

**Comment by caglarsu** [comment]
> Contacting with Tesla fix our problem. They added it manually behalf of us.

I sent an email Tesla more than a week ago and there is no respond :( We are completely stuck and cannot move forward. Very sad story for us.

**Comment by sethterashima** [comment]
@martinzitka @sgryphon When users pair the virtual key using the QR code from tesla.com/_ak/example.com, the mobile app will try to fetch your public key from example.com and verify it matches the one Tesla has on record. This way the app doesn't need to trust Tesla's servers for public key distribution. But it does mean that your website needs to be reachable by the mobile app, and so you may want to consider making your site IPv4 accessible.

**Comment by sgryphon** [comment]
Okay, so if there is a mobile user whose carrier doesn't support modern connections then we may want to add some legacy support, such as an IPv4 reverse proxy (I use Cloudflare for some of my domains for this). I have a limited user base where I know there is no one in this category, but it may be relevant for some people to consider.

However Tesla may also want to consider that they fully support the current internet version, IPv6.

Most mobile carriers support IPv6, in fact some such my local carrier, Telstra in Australia, have moved to IPv6-only https://www.sidn.nl/en/news-and-blogs/australias-telstra-switches-mobile-users-to-ipv6-only 

There are probably more IPv6-only carriers than IPv4-only carriers, at least for consumer devices. (They do have legacy support using XLAT464 with carrier-grade NAT)

Also, if you have an Apple IOS app, then all apps submitted to the App Store since 01 June *2016* (8 years ago) must support IPv6-only networks -- if your app does not support IPv6-only networks, then it may be in breach. https://developer.apple.com/support/ipv6/

It is probably good to note / give a warning if you detect a site only supports IPv6 about the way the app works if a user only has IPv4, but you should also fully support modern IPv6.

**Comment by sgryphon** [comment]
Out of interest I also just checked my local LAN for both my Model 3 and Tesla Wall Charger. I run a dual-stack network.

The devices do not request a DHCPv6 lease, and do not respond to an EUI-64 auto-configured address, although they do respond to IPv4 pings.

For the car:

![image](https://github.com/user-attachments/assets/f451d8a1-3107-4f20-b180-94e6886e4392)

For the charger:

![image](https://github.com/user-attachments/assets/d4c134d5-2c4a-485a-9db5-15cc5fae50ac)

This is not good to not support modern internet and only support legacy protocols. 

There are also some (not many) ISPs that provide IPv6-only (primarily due to the high costs involved in legacy IPv4), so those customers may not be able to use your products if they are IPv4 only.

**Comment by djballowe** [comment]
@handya 

I have the exact same issue with a domain that worked a month ago. did this just sort itself out over time?

**Comment by handya** [comment]
> @handya
> 
> I have the exact same issue with a domain that worked a month ago. did this just sort itself out over time?

Yes, I tried again a few days later and it started working again 🤷‍♂️

**Comment by orosys** [comment]
If the domain is not registered on the Tesla developer site, try clicking the “Next” or “Update” button repeatedly. It should eventually proceed. In my case, please attempt this under the assumption that you can successfully access .well-known/appspecific/com.tesla.3p.public-key.pem.

**Comment by acanturgut** [comment]
> Any more help on this? I'm attempting to register the domain `https://www.myteslapower.co.uk/`
> 
> ![image](https://private-user-images.githubusercontent.com/57369759/367942295-dd62e496-af85-44aa-a4db-a118d969fb15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY1MjU3OTcsIm5iZiI6MTcyNjUyNTQ5NywicGF0aCI6Ii81NzM2OTc1OS8zNjc5NDIyOTUtZGQ2MmU0OTYtYWY4NS00NGFhLWE0ZGItYTExOGQ5NjlmYjE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDIyMjQ1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzYjkyNjRhNGMzNDUwNzQ1MTNjYTdmNDY4MjllMzQxM2JhOTc0ZjcyMGMyY2ZkYzNiOWQ5N2Q0YWUxZmY3MWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vpzsdMKqGstvHCa6Ap1or3egwREoqaKzI7lgrHUHNvc)
> 
> I meet the so-called "requirements", e.g CN matching and a valid certificate.
> 
> If anyone could help that would be much appreciated as none of the above solutions worked and I'm kind of stuck. I already contacted Tesla Fleet API support but no respond regarding the actual issue.

Your domain looks fine, better to nagging tesla costumer service. They will add the domain manually eventually. Better to append localhost or other domains which is you desired to use in the future as well on your email.

**Comment by JjyKsi** [comment]
Added ionkeep.app couple days ago and worked fine. Came back today to add some testing urls and all of sudden it says that ionkeep.app isn't ok? 

Using Let's Encrypt, CN seems to match, the certificate is valid etc. Any idea what's going on?

**Comment by Den-88** [comment]
I am experiencing issues with my websites:

https://teslatrack.tech/
https://fleet.teslatrack.tech/
https://telemetry.teslatrack.tech/
The main site, https://teslatrack.tech/, worked fine after the initial setup, but after a couple of days, it stopped functioning. I have been unable to get any of these sites to work properly.

Could you please check what might be causing the problem?

<img width="1743" alt="image" src="https://github.com/user-attachments/assets/8fd697c3-3a1b-4a1f-8293-2680445835da">

**Comment by getget1980** [comment]
Hello experiencing same problems with .fr

**Comment by ThomasAlxDmy** [comment]
@Den-88 try to remove Tesla from DNS name, there's logic on ourside to ensure customer don't get confused this is a 3rd party.

**Comment by jman-bunker** [comment]
I've texted a bit with Tesla. They also suggested that I changed my dns from tesla.<something>.synology.me. So I did. Changed to car.<something>.synology.me and it still did not work. But Tesla just chose to ignore me from that point... Pretty horrible customer service imho

**Comment by Jopand** [comment]
I've texted a bit with Tesla. They also suggested that I changed my DNS from tesla.<something>.synology.me. So I did. Changed to car.<something>.synology.me and it still did not work. But Tesla just chose to ignore me from that point... Pretty horrible customer service imho

**Comment by jbanyer** [comment]
> @Den-88 try to remove Tesla from DNS name, there's logic on ourside to ensure customer don't get confused this is a 3rd party.

@ThomasAlxDmy can we get some more info about this please?

Our registered domain is tesla.chargehq.net and we have many vehicles paired to this domain. When we set it up there was no restriction preventing the domain including the word "tesla" and it's been working fine for many months.

Currently (4th Dec) clicking our [pairing link](https://tesla.com/_ak/tesla.chargehq.net/) on the mobile results in this error:

![image](https://github.com/user-attachments/assets/79a1611f-a797-48d4-a01f-2237fdd3656e)

This is preventing us from onboarding new users. A reply would be appreciated. Thanks!

**Comment by corsair** [comment]
> @Den-88 try to remove Tesla from DNS name, there's logic on ourside to ensure customer don't get confused this is a 3rd party.

@ThomasAlxDmy I also have concerns about this recent change. We discussed this topic over a year ago with Tesla in October 2023. The decision was then made to revise the validation to allow the term "tesla" from now on.

<img width="412" alt="image" src="https://github.com/user-attachments/assets/b051c4fe-8c06-4c54-b276-264d004f1c21">

For many services, such as my own, "removing Tesla" immediately would result in vehicle owner concerns over phishing or vehicle safety when a vehicle key is present that does not precisely match the root domain of the service they have authorized. When this validation briefly existed last year, we polled our community, and 70% replied that they would feel uncomfortable adding a vehicle key that did not match the root domain.

If a customer signed up for "Teslascope.com" and saw "ThirdPartyScope.com" had access to their location, they would assume my Tesla Account was compromised and contact Tesla out of concern.

I hope this change will be reconsidered or modified to at least exclude the root TLD; otherwise, this opens up more risks to Tesla customers and vehicle owners.

**Comment by jbanyer** [comment]
> Currently (4th Dec) clicking our pairing link https://tesla.com/_ak/tesla.chargehq.net/ on the mobile results in this error:
> 
> ![image](https://private-user-images.githubusercontent.com/936491/392155365-79a1611f-a797-48d4-a01f-2237fdd3656e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0NjQ2NzIsIm5iZiI6MTczMzQ2NDM3MiwicGF0aCI6Ii85MzY0OTEvMzkyMTU1MzY1LTc5YTE2MTFmLWE3OTctNDhkNC1hMDFmLTIyMzdmZGQzNjU2ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwNlQwNTUyNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NTc5M2UwMzAwMTFjYjExMjE2OWFjZmZkNzliOWI2ZWE1ZmE4MGQzNDA0MGJiZWQxNjI2MmNlYzNmMjMyYzJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.p6J1hpwabeUb5ztylDo_CinwJL_n5woztCCaQZNVSjo)

Actually this error was caused by the trailing slash "/" (thanks @patrickdemers6). The links work fine if the slash is removed. So it appears that the domains can include the word "tesla" - at least that's true for existing domains.

Apologies for the misleading error report.

**Comment by michaelehlers** [comment]
> It is NOT working on valid *.synology.me, *.mywire.org and *.duckdns.org domains 🤔



> I've texted a bit with Tesla. They also suggested that I changed my DNS from tesla..synology.me. So I did. Changed to car..synology.me and it still did not work. But Tesla just chose to ignore me from that point... Pretty horrible customer service imho

If you happened to get a response would be great to share @Jopand . Stuck with exactly the same problem. synology.me, let's encryp with valid cert and CN, not 'tesla' in the url.... :/

**Comment by moinal** [comment]
Hi,
Can't add kilinka.com or www.kilinka.com
When i test the domain, everything seems to be ok.
https://www.ssllabs.com/ssltest/analyze.html?d=kilinka.com
Please help
